### PR TITLE
Missions

### PR DESCRIPTION
**Pull request by PeeGee85Spil**
- Added missions to SDK, example app and unity bridge.
- Can be tested via the missions menu in the sample app, missions/containers can be started/completed/collected.
- Refactored updateInventoryWithBundle (android did the same), can probably refactor more PlayerDataController methods to make use of the new setPlayerDataWithWallet and setPlayerDataWithInventory methods, instead of repeating the same code.

### CODE REVIEW

File: SampleApp/Src/Missions/MissionsViewController.m [hunk 1]
```diff
@@ -0,0 +1,471 @@
...
+        // ...
+        
+        if(missionProgress.status != nil && [SpilMissionsHandler statusFromNSString:missionProgress.status] == IN_PROGRESS) {
+            if(
+               (mission.rewards != nil && mission.rewards.count > 0) ||
+               (mission.gameUnlocks != nil && mission.gameUnlocks.count > 0) ||
+               (mission.missionUnlocks != nil && mission.missionUnlocks.count > 0)
+            ) {
+                missionProgress.status = [SpilMissionsHandler statusToUnityString:PENDING_COLLECTION];
+                // TODO: A bit odd maybe that devs would have to set this themselves?
```
Comment (FrankSpil) [inline SampleApp/Src/Missions/MissionsViewController.m]: This is correct, todo can be removed

File: SampleApp/Src/Missions/MissionsViewController.m [hunk 2]
```diff
@@ -0,0 +1,471 @@
...
+            } else {
+                missionProgress.status = [SpilMissionsHandler statusToUnityString:COMPLETED];
+            }
+            
+            // Fill progress with any final values
+            // ...
+        
+            // TODO: Change lastcompleted on completion or on collection?
```
Comment (FrankSpil) [inline SampleApp/Src/Missions/MissionsViewController.m]: This has to be done on completion!

File: SampleApp/Src/Missions/MissionsViewController.m [hunk 3]
```diff
@@ -0,0 +1,471 @@
...
+                [missionsToReward addObject:missionProgress];
+                [Spil claimMissionsReward:missionsToReward withReason:@"Test App" withReasonDetails:@"Reward Claiming" withLocation:@"Mission Completion" withTransactionId:nil];
+            }
+            
+            // Fill progress with any final values
+            // ...
+            
+            missionProgress.status = [SpilMissionsHandler statusToUnityString:COMPLETED];
+            
+            // TODO: Change lastcompleted on completion or on collection?
```
Comment (FrankSpil) [inline SampleApp/Src/Missions/MissionsViewController.m]: This has to be done on completion!

File: SampleApp/Src/Missions/MissionsViewController.m [hunk 4]
```diff
@@ -0,0 +1,471 @@
...
+            ) {
+                containerProgress.status = [SpilMissionsHandler statusToUnityString:PENDING_COLLECTION];
+            } else {
+                containerProgress.status = [SpilMissionsHandler statusToUnityString:COMPLETED];
+            }
+            
+            // Fill progress with any final values
+            // ...
+            
+            // TODO: Change lastcompleted on completion or on collection?
```
Comment (FrankSpil) [inline SampleApp/Src/Missions/MissionsViewController.m]: This has to be done on completion!

File: SampleApp/Src/Missions/MissionsViewController.m [hunk 5]
```diff
@@ -0,0 +1,471 @@
...
+                [containersToReward addObject:containerProgress];
+                [Spil claimContainersReward:containersToReward withReason:@"Test App" withReasonDetails:@"Reward Claiming" withLocation:@"Mission Completion" withTransactionId:nil];
+            }
+            
+            containerProgress.status = [SpilMissionsHandler statusToUnityString:COMPLETED];
+            
+            // Fill progress with any final values
+            // ...
+            
+            // TODO: Change lastcompleted on completion or on collection?
```
Comment (FrankSpil) [inline SampleApp/Src/Missions/MissionsViewController.m]: This has to be done on completion!

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 6]
```diff
@@ -0,0 +1,958 @@
...
+            [progressToBeSentJSON addObject:[currentProgress toBackEndJSONObject]];
+        }
+    }
+    
+    if(progressToBeSent.count == 0){
+        return;
+    }
+    
+    /* TODO: Hmmmmmkay. What's going on here? Align with Seb/Frank.
+     userData = UserDataManager.getInstance(context).updateUserDataVersion(userData);
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: [[SpilUserHandler sharedInstance] increaseDeviceVersion]; has to be called here, it does the same thing as the 2 android methods.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 7]
```diff
@@ -0,0 +1,958 @@
...
+            [progressToBeSentJSON addObject:[currentProgress toBackEndJSONObject]];
+        }
+    }
+    
+    if (progressToBeSent.count == 0) {
+        return;
+    }
+    
+    /* TODO: Hmmmmmkay. What's going on here? Align with Seb/Frank.
+     userData = UserDataManager.getInstance(context).updateUserDataVersion(userData);
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: [[SpilUserHandler sharedInstance] increaseDeviceVersion]; has to be called here, it does the same thing as the 2 android methods.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 8]
```diff
@@ -0,0 +1,958 @@
...
+                [itemObjects addObject:[[PlayerDataController sharedInstance] updateItemReward:userProfile withItemId:reward.id withItemAmount:reward.amount withPromotion:nil withPerkItems:nil]];
+            }
+        }
+        
+        //[updatedUserData.claimedContainers addObject:containerProgress.containerId];
+    }
+    
+    // Updates UserData Version and Meta
+    /* TODO: Hmmmmmkay. What's going on here? Align with Seb/Frank.
+     userData = UserDataManager.getInstance(context).updateUserDataVersion(userData);
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: [[SpilUserHandler sharedInstance] increaseDeviceVersion]; has to be called here, it does the same thing as the 2 android methods.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 9]
```diff
@@ -0,0 +1,958 @@
...
+
+    UserProfile* userProfile = [[PlayerDataController sharedInstance] getUserProfile];
+    if (userProfile == nil) {
+        // Inform the listener
+        NSString *errorMessage = [[SpilError LoadFailed:@"claimContainersReward"] toJson];
+        if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+            [[Spil sharedInstance].delegate userDataError:errorMessage];
+        }
+        [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+        // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo can be removed, looks fine.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 10]
```diff
@@ -0,0 +1,958 @@
...
+    }
+    
+    if (containerProgressList == nil) {
+        // Inform the listener
+        NSString *errorMessage = [[SpilError ContainerOperation] toJson];
+        if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+            [[Spil sharedInstance].delegate userDataError:errorMessage];
+        }
+        [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+        // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo can be removed, looks fine.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 11]
```diff
@@ -0,0 +1,958 @@
...
+    
+    for (ContainerProgress* containerProgress in containerProgressList) {
+        if ([SpilMissionsHandler statusFromNSString:containerProgress.status] != PENDING_COLLECTION) {
+            // Inform the listener
+            NSString *errorMessage = [[SpilError ContainerOperation] toJson];
+            if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+                [[Spil sharedInstance].delegate userDataError:errorMessage];
+            }
+            [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+            // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo can be removed, looks fine.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 12]
```diff
@@ -0,0 +1,958 @@
...
+     userData = UserDataManager.getInstance(context).updateUserDataMeta(userData);
+    */
+    
+    //Saves updated user data
+    [[PlayerDataController sharedInstance] updateUserProfile:userProfile];
+    
+    //TODO: Android seems to be adding some gacha info from the userprofile to the updateddata, is this needed for iOS as well?
+    /*
+    for (PlayerItem* playerItem in userData.inventory.items) {
+        // 1 == true
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Not needed for iOS

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 13]
```diff
@@ -0,0 +1,958 @@
...
+
+    UserProfile* userProfile = [[PlayerDataController sharedInstance] getUserProfile];
+    if (userProfile == nil) {
+        // Inform the listener
+        NSString *errorMessage = [[SpilError LoadFailed:@"claimMissionsReward"] toJson];
+        if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+            [[Spil sharedInstance].delegate userDataError:errorMessage];
+        }
+        [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+        // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo, can be removed, checked and looks fine

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 14]
```diff
@@ -0,0 +1,958 @@
...
+    }
+    
+    if (missionProgressList == nil) {
+        // Inform the listener
+        NSString *errorMessage = [[SpilError ContainerOperation] toJson];
+        if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+            [[Spil sharedInstance].delegate userDataError:errorMessage];
+        }
+        [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+        // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo, can be removed, checked and looks fine

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 15]
```diff
@@ -0,0 +1,958 @@
...
+    
+    for (MissionProgress* missionProgress in missionProgressList) {
+        if ([SpilMissionsHandler statusFromNSString:missionProgress.status] != PENDING_COLLECTION) {
+            // Inform the listener
+            NSString *errorMessage = [[SpilError ContainerOperation] toJson];
+            if ([[Spil sharedInstance].delegate respondsToSelector:@selector(userDataError:)]) {
+                [[Spil sharedInstance].delegate userDataError:errorMessage];
+            }
+            [Spil sendMessage:@"UserDataError" toObject:@"SpilSDK" withString:errorMessage];
+            // TODO: Why not send to unity here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Todo, can be removed, checked and looks fine

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 16]
```diff
@@ -0,0 +1,958 @@
...
+     */
+    
+    //Saves updated user data
+    [[PlayerDataController sharedInstance] updateUserProfile:userProfile];
+    
+    //TODO: Android seems to be adding some gacha info from the userprofile to the updateddata, is this needed for iOS as well?
+    /*
+     for (PlayerItem* playerItem in userData.inventory.items) {
+     // 1 == true
+     if (playerItem.isGacha == 1 && playerItem.delta != 0) {
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: This is not needed for iOS.

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 17]
```diff
@@ -0,0 +1,958 @@
...
+                [itemObjects addObject:[[PlayerDataController sharedInstance] updateItemReward:userProfile withItemId:reward.id withItemAmount:reward.amount withPromotion:nil withPerkItems:nil]];
+            }
+        }
+        
+        //updatedUserData.claimedMissions.add(missionProgress.getMissionId());
+    }
+    
+    // Updates UserData Version and Meta
+    /* TODO: Hmmmmmkay. What's going on here? Align with Seb/Frank.
+     userData = UserDataManager.getInstance(context).updateUserDataVersion(userData);
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: [[SpilUserHandler sharedInstance] increaseDeviceVersion]; can be called, it does the same thing as the 2 android calls

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 18]
```diff
@@ -0,0 +1,958 @@
...
+    }
+    
+    userData.missionData.containerProgressList = [NSMutableArray array];
+    userData.missionData.missionProgressList = [NSMutableArray array];
+    
+    [[PlayerDataController sharedInstance] updateUserProfile:userData];
+    
+    NSDictionary* userProfile = [userData toJSONObject];
+    
+    // TODO: Reset is not supported by the backend atm.
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: I don't think there will be an event for this ever, the todo can be removed

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 19]
```diff
@@ -0,0 +1,958 @@
...
+       @"missionProgress" : userProfile[@"missionData"][@"missionProgress"],
+       @"reason" : @"Reset",
+       @"reasonDetails" : @"",
+       @"location" : @"",
+       @"transactionId" : @""
+    };
+    [[SpilEventTracker sharedInstance] trackEvent:@"updateUserMissionData" withParameters:data onResponse:^(NSDictionary *response) {
+    }];
+    
+    // TODO: Why not fire MissionDataUpdated callback here?
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: Is this todo still valid?

File: SpilSDK/Src/PlayerData/PlayerDataController.m [hunk 20]
```diff
@@ -391,6 +400,7 @@ -(void)setPlayerDataWithWallet:(Wallet*)updatedWallet withInventory:(Inventory*)
     [updatedData setObject:uniqueItemObjects forKey:@"uniqueItems"];
     NSArray* currencyObjects = [userProfile.wallet getCurrenciesJSONArray];
     [updatedData setObject:currencyObjects forKey:@"currencies"];
+    // TODO: Add missions here as well?
```
Comment (FrankSpil) [inline SpilSDK/Src/PlayerData/PlayerDataController.m]: I don't think it's needed here

File: SpilSDK/Src/PlayerData/PlayerDataController.m [hunk 21]
```diff
@@ -1079,6 +996,134 @@ -(void)updateInventoryWithBundle:(int)bundleId withReason:(NSString*)reason with
     }
 }
 
+// TODO: More methods in this class could use this method, instead of repeating the same code.
```
Comment (FrankSpil) [inline SpilSDK/Src/PlayerData/PlayerDataController.m]: Nice util methods!

File: SpilSDK/Src/PlayerData/PlayerDataController.m [hunk 22]
```diff
@@ -1079,6 +996,134 @@ -(void)updateInventoryWithBundle:(int)bundleId withReason:(NSString*)reason with
...
+        @"delta":[NSNumber numberWithInteger:playerCurrency.delta],
+        @"overflow":[NSNumber numberWithInteger:playerCurrency.overflow],
+        @"initialValue":[NSNumber numberWithInteger:playerCurrency.initialValue],
+        @"limit":[NSNumber numberWithInteger:playerCurrency.limit],
+        @"name":playerCurrency.name,
+        @"type":[NSNumber numberWithInteger:playerCurrency.type]
+    };
+}
+
+// TODO: More methods in this class could use this method, instead of repeating the same code.
```
Comment (FrankSpil) [inline SpilSDK/Src/PlayerData/PlayerDataController.m]: Nice util methods!

File: SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m [hunk 23]
```diff
@@ -0,0 +1,958 @@
...
+            // Not a unique item
+            [filteredItemObjects addObject:playerItem];
+        }
+    }
+    
+    // Send notification
+    NSMutableDictionary *updatedData = [NSMutableDictionary dictionary];
+    [updatedData setObject:filteredItemObjects forKey:@"items"];
+    [updatedData setObject:currencyObjects forKey:@"currencies"];
+    NSDictionary *userInfo = @{@"event" : @"playerDataUpdated", @"reason" : reason, @"updatedData" : updatedData};
```
Comment (FrankSpil) [inline SpilSDK/Src/MissionsHandler/SpilMissionsHandler.m]: claimedContainers & claimedMissions should be added

File: SpilSDK/Src/Util/SpilError.m [hunk 24]
```diff
@@ -108,6 +108,22 @@ +(SpilError*)ConfigServerError {
     return [[SpilError alloc] initWithId:19 name:@"ConfigServerError" message:@"Error retrieving config"];
 }
 
++(SpilError*)MissionConfigurationServerError {
+    return [[SpilError alloc] initWithId:19 name:@"MissionConfigurationServerError" message:@"Error retrieving mission configuration."];
```
[the remark this answers is not in the document]
Comment (PeeGee85Spil) [reply SpilSDK/Src/Util/SpilError.m]: Has been fixed, only one error with id 19 now.

File: SpilSDK/Src/MissionsHandler/New Group/ContainerProgress.h [hunk 25]
```diff
@@ -0,0 +1,24 @@
+//
```
[the remark this answers is not in the document]
Comment (PeeGee85Spil) [reply SpilSDK/Src/MissionsHandler/New Group/ContainerProgress.h]: Renamed to models

**Pull request merged**